src/bytes.rs

```diff
@@ -2260,6 +2345,94 @@ unsafe impl Sync for Inner {}
  *
  */
 
+impl Inner2 {
+    fn into_vec(self) -> Vec<u8> {
+        let kind = self.inner.kind();
+        if kind == KIND_VEC {
+            let r = unsafe { Vec::from_raw_parts(self.inner.ptr, self.inner.len, self.inner.cap) };
+
+            // no need to call the destructor, we took the data
+            mem::forget(self);
+
+            return r;
+        }
+
+        if kind == KIND_ARC {
+            let arc = self.arc.load(Relaxed);
+            if unsafe { (*arc).is_unique() } {
+                unsafe {
+                    // copy variables because we forget `self` before we need them
+                    let self_ptr = self.ptr;
+                    let self_len = self.len;
+
+                    let shared: Box<Shared> = mem::transmute(arc);
+
+                    // We took ownership of both `Shared` and `Vec`,
+                    // no need to call destructor.
+                    //
+                    // Call `forget` early to make sure there won't be double free
+                    // in case code below panics (which shouldn't).
+                    mem::forget(self);
+
+                    let shared: Shared = *shared;
+                    let mut r = shared.vec;
+                    // offset of data within vec
+                    let offset = self_ptr as usize - r.as_slice().as_ptr() as usize;
+                    r.truncate(offset + self_len);
+                    r.drain(..offset);
+                    return r;
+                }
+            }
+        }
+
+        // default impl
+        self.as_ref().to_owned()
+    }
+
+    // `unsafe` since it must be called with a uniquely owned buffer (`BytesMut`)
+    unsafe fn into_vec_mut(self) -> Vec<u8> {
+        let kind = self.inner.kind();
+        if kind == KIND_VEC {
+            let r = Vec::from_raw_parts(self.inner.ptr, self.inner.len, self.inner.cap);
+
+            // no need to call the destructor, we took the data
+            mem::forget(self);
+
+            return r;
+        }
+
+        if kind == KIND_ARC {
```

Comment:
If I'm reading this correctly, this function assumes that a `BytesMut` handle has a unique `Arc<Shared>` (no other clones). However, that is not the case. There can be many outstanding handles to the same `Arc<Shared>`. `BytesMut` only guarantees that it has exclusive access to a window of the shared buffer. So, it can mutate the bytes in its window, but it does not have exclusive ownership of the buffer.

My guess is this function should be deleted in favor of just always calling `Inner2::into_vec`.